assets/js/modules/analytics-4/datastore/report.js

```diff
@@ -0,0 +1,175 @@
+			'Options for Analytics 4 report must be an object.'
+		);
+		invariant(
+			isValidDateRange( options ),
+			'Either date range or start/end dates must be provided for Analytics 4 report.'
+		);
+
+		const { metrics, dimensions, dimensionFilters, orderby } =
+			normalizeReportOptions( options );
+
+		invariant(
+			metrics.length,
+			'Requests must specify at least one metric for an Analytics 4 report.'
+		);
+		invariant(
+			isValidMetrics( metrics ),
+			'Metrics for an Analytics 4 report must be either a string, an array of strings, an object, an array of objects or a mix of strings and objects. If an object is used, it must have "name" property and optionally a "expression" property.'
+		);
+
+		if ( dimensions ) {
+			invariant(
+				isValidDimensions( dimensions ),
+				'Dimensions for an Analytics 4 report must be either a string, an array of strings, an object, an array of objects or a mix of strings and objects. If an object is used, it must have "name" property.'
+			);
+		}
+
+		if ( dimensionFilters ) {
+			invariant(
+				isValidDimensionFilters( dimensionFilters ),
+				'Dimension filters must be a map of dimension names as keys and dimension values as values.'
+			);
+		}
+
+		if ( orderby ) {
+			invariant(
+				isValidOrders( orderby ),
+				'Orders for an Analytics report must be either an object or an array of objects where each object should have "fieldName" and "sortOrder" properties.'
+			);
+		}
+	},
+} );
+
+const baseInitialState = {
+	reports: {},
+};
+
+const baseResolvers = {
+	*getReport( options = {} ) {
+		const registry = yield Data.commonActions.getRegistry();
+		const existingReport = registry
+			.select( MODULES_ANALYTICS_4 )
+			.getReport( options );
+
+		// If there is already a report loaded in state, consider it fulfilled
+		// and don't make an API request.
+		if ( existingReport ) {
+			return;
+		}
+
+		yield fetchGetReportStore.actions.fetchGetReport( options );
+	},
+};
+
+const baseSelectors = {
+	/**
+	 * Gets an Analytics report for the given options.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object}         state                       Data store's state.
+	 * @param {Object}         options                     Options for generating the report.
+	 * @param {string}         options.startDate           Required, unless dateRange is provided. Start date to query report data for as YYYY-mm-dd.
+	 * @param {string}         options.endDate             Required, unless dateRange is provided. End date to query report data for as YYYY-mm-dd.
+	 * @param {string}         options.dateRange           Required, alternative to startDate and endDate. A date range string such as 'last-28-days'.
+	 * @param {Array.<string>} options.metrics             Required. List of metrics to query.
+	 * @param {boolean}        [options.compareDateRanges] Optional. Only relevant with dateRange. Default false.
+	 * @param {boolean}        [options.multiDateRange]    Optional. Only relevant with dateRange. Default false.
```

Comment:
A few of these were not implemented (as they've since been essentially deprecated by `dateRange` and `compare(Start|End)Date` options. Specifically `dateRange`, `compareDateRanges`, and `multiDateRange` should be removed here.